### DISCUSSION TITLE
ArgoCD dexserver configuration fails with ldapserver authentication

### DISCUSSION THREAD

**Discussion by divya-prakash2** [argoproj/argo-cd#7693] [Q&A]
Describe the bug
Need help in configuring argocd with ldap authentication

To Reproduce
argocd-cm has been edited to have the dex.config information as follows
dex.config: |-
connectors:
- type: ldap
name: myad.local
id: ldap
config:
# Ldap server address
host: 172.28.8.3:10389
insecureNoSSL: true
insecureSkipVerify: true
# Variable name stores ldap bindDN in argocd-secret
bindDN: ou=admin,dc=sdomain,dc=com
# Variable name stores ldap bind password in argocd-secret
bindPW: *****
usernamePrompt: ""
# Ldap user serch attributes
userSearch:
baseDN: ou=people,dc=sdomain,dc=com
filter: "(objectClass=inetOrgPerson)"
username: uid
idAttr: uid
emailAttr: mail
nameAttr: displayName
# Ldap group serch attributes
groupSearch:
baseDN: ou=people,dc=sdomain,dc=com
filter: "(objectClass=inetOrgPerson)"
userAttr: uid
groupAttr: uid
nameAttr: displayName

data.url has been provided the argocd URL

Clicking on argoCD login via myad.local gives the following error (503 Service Unavailable )
Failed to query provider "https://172.28.8.9/api/dex": 503 Service Unavailable:

<TITLE>Network Error</TITLE>
argocd-dex-server logs
lockmaker01:/opt/glcg/certificates # kubectl logs pod/argocd-dex-server-6d59765b55-8tqm2 -n argocd
time="2021-11-08T14:28:58Z" level=info msg="Starting configmap/secret informers"
time="2021-11-08T14:28:58Z" level=info msg="Configmap/secret informer synced"
time="2021-11-08T14:28:58Z" level=info msg="Loading TLS configuration from secret argocd/argocd-server-tls"
time="2021-11-08T14:28:58Z" level=info msg="0xc000706960 subscribed to settings updates"
time="2021-11-08T14:28:58Z" level=info msg="config issuer: https://172.28.8.9/api/dex"
time="2021-11-08T14:28:58Z" level=info msg="config storage: memory"
time="2021-11-08T14:28:58Z" level=info msg="config static client: Argo CD"
time="2021-11-08T14:28:58Z" level=info msg="config static client: Argo CD CLI"
time="2021-11-08T14:28:58Z" level=info msg="config connector: ldap"
time="2021-11-08T14:28:58Z" level=info msg="config skipping approval screen"
time="2021-11-08T14:28:58Z" level=info msg="keys expired, rotating"
time="2021-11-08T14:28:58Z" level=info msg="keys rotated, next rotation: 2021-11-08 20:28:58.401423999 +0000 UTC"
time="2021-11-08T14:28:58Z" level=info msg="listening (http/telemetry) on 0.0.0.0:5558"
time="2021-11-08T14:28:58Z" level=info msg="listening (http) on 0.0.0.0:5556"
time="2021-11-08T14:28:58Z" level=info msg="listening (grpc) on 0.0.0.0:5557"

argocd-server logs
time="2021-11-08T14:29:45Z" level=info msg="Initializing OIDC provider (issuer: https://172.28.8.9/api/dex)"
time="2021-11-08T14:30:48Z" level=info msg="Initializing OIDC provider (issuer: https://172.28.8.9/api/dex)"
time="2021-11-08T14:38:56Z" level=info msg="Alloc=10251 TotalAlloc=31662 Sys=74321 NumGC=12 Goroutines=158"
time="2021-11-08T14:48:56Z" level=info msg="Alloc=10204 TotalAlloc=35580 Sys=74321 NumGC=17 Goroutines=156"
time="2021-11-08T14:58:56Z" level=info msg="Alloc=10156 TotalAlloc=39296 Sys=74321 NumGC=22 Goroutines=156"
time="2021-11-08T15:08:56Z" level=info msg="Alloc=10231 TotalAlloc=43109 Sys=74321 NumGC=27 Goroutines=156"
time="2021-11-08T15:18:56Z" level=info msg="Alloc=10183 TotalAlloc=46941 Sys=74321 NumGC=32 Goroutines=156"
time="2021-11-08T15:28:56Z" level=info msg="Alloc=10182 TotalAlloc=50744 Sys=74321 NumGC=37 Goroutines=156"
time="2021-11-08T15:38:56Z" level=info msg="Alloc=10123 TotalAlloc=54315 Sys=74321 NumGC=42 Goroutines=156"
time="2021-11-08T15:48:56Z" level=info msg="Alloc=10243 TotalAlloc=58122 Sys=74321 NumGC=47 Goroutines=156"
time="2021-11-08T15:58:49Z" level=info msg="Initializing OIDC provider (issuer: https://172.28.8.9/api/dex)"
time="2021-11-08T15:58:56Z" level=info msg="Alloc=10300 TotalAlloc=61994 Sys=74321 NumGC=52 Goroutines=157"
time="2021-11-08T15:59:03Z" level=info msg="Initializing OIDC provider (issuer: https://172.28.8.9/api/dex)"
time="2021-11-08T16:00:14Z" level=info msg="Initializing OIDC provider (issuer: https://172.28.8.9/api/dex)"

Questions:
What needs to be provided in data.url ?
Are there any special configurations that needs to be performed on argocd-dex-server pod ?
Where do we configure dex server endpoint /api/dex/callback ?

**Comment by leoluz** [comment]
> What needs to be provided in data.url ?

Argo CD's externally facing base URL

> Are there any special configurations that needs to be performed on argocd-dex-server pod ?

All configurations to dex server is provided by the argocd-cm

> Where do we configure dex server endpoint /api/dex/callback ?

This configuration should go in your LDAP provider

**Comment by divya-prakash2** [comment]
@leoluz , Firstly thanks for your response

Does ArgoCD integration with LDAP Server need to have .data.url parameter in configmap argocd-cm?
Do you see any issues with the dex.config added here ? 
Clicking in Ldap login is giving 503 service unavailable if .data.url is provided to argocd base URL ?
 If .data.url is not provided , it keep redirecting the same login page . 

 
![image](https://user-images.githubusercontent.com/55496734/141161399-6a1e2619-2b0c-4533-8f10-9ad6283cd6dd.png)

**Comment by leoluz** [comment]
`data.url` is the public facing URL that points to your ArgoCD server. This needs to be accessible by your LDAP server.
One strange thing from your error message bellow is that I see the IP `172.28.8.9`. I was expecting to have it defined with default value of the Dex service which would be `http://argocd-dex-server:5556`: 
```
Failed to query provider "https://172.28.8.9/api/dex": 503 Service Unavailable:
```
Are you reconfiguring the Dex server (`ARGOCD_SERVER_DEX_SERVER`) to point to this IP?

**Comment by divya-prakash2** [comment]
@leoluz 
Whatever provided in .data.url is getting populated into the error which you pointed out 
Failed to query provider "https://172.28.8.5/api/dex": 503 Service Unavailable:

172.28.8.5 is the public facing URL for ArgoCD server 

Dex server logs doesnt show any error 
lockmaker01:/opt/gateway/yaml # kubectl logs pod/argocd-dex-server-6d59765b55-m6bcm -n argocd
time="2021-11-10T09:56:45Z" level=info msg="Starting configmap/secret informers"
time="2021-11-10T09:56:45Z" level=info msg="Configmap/secret informer synced"
time="2021-11-10T09:56:45Z" level=info msg="0xc0007bc360 subscribed to settings updates"
time="2021-11-10T09:56:45Z" level=info msg="dex is not configured"
time="2021-11-10T09:57:23Z" level=info msg="Notifying 1 settings subscribers: [0xc0007bc360]"
time="2021-11-10T09:57:23Z" level=info msg="dex config modified. restarting dex"
**time="2021-11-10T09:57:23Z" level=info msg="config issuer: https://172.28.8.5/api/dex"**
time="2021-11-10T09:57:23Z" level=info msg="config storage: memory"
time="2021-11-10T09:57:23Z" level=info msg="config static client: Argo CD"
time="2021-11-10T09:57:23Z" level=info msg="config static client: Argo CD CLI"
time="2021-11-10T09:57:23Z" level=info msg="config connector: ldap"
time="2021-11-10T09:57:23Z" level=info msg="config skipping approval screen"
time="2021-11-10T09:57:23Z" level=info msg="keys expired, rotating"
time="2021-11-10T09:57:23Z" level=info msg="keys rotated, next rotation: 2021-11-10 15:57:23.954978598 +0000 UTC"
time="2021-11-10T09:57:23Z" level=info msg="listening (http/telemetry) on 0.0.0.0:5558"
time="2021-11-10T09:57:23Z" level=info msg="listening (http) on 0.0.0.0:5556"
time="2021-11-10T09:57:23Z" level=info msg="listening (grpc) on 0.0.0.0:5557"
time="2021-11-10T09:57:24Z" level=info msg="Notifying 1 settings subscribers: [0xc0007bc360]"
time="2021-11-10T09:57:24Z" level=info msg="dex config unmodified"
time="2021-11-10T09:57:24Z" level=info msg="Notifying 1 settings subscribers: [0xc0007bc360]"
time="2021-11-10T09:57:24Z" level=info msg="dex config unmodified"
time="2021-11-10T15:57:24Z" level=info msg="keys expired, rotating"
time="2021-11-10T15:57:24Z" level=info msg="keys rotated, next rotation: 2021-11-10 21:57:24.352839712 +0000 UTC"
time="2021-11-10T21:57:24Z" level=info msg="keys expired, rotating"
time="2021-11-10T21:57:24Z" level=info msg="keys rotated, next rotation: 2021-11-11 03:57:24.562836256 +0000 UTC"
time="2021-11-11T03:57:24Z" level=info msg="keys expired, rotating"
time="2021-11-11T03:57:25Z" level=info msg="keys rotated, next rotation: 2021-11-11 09:57:25.20468318 +0000 UTC"
time="2021-11-11T09:57:25Z" level=info msg="keys expired, rotating"
time="2021-11-11T09:57:25Z" level=info msg="keys rotated, next rotation: 2021-11-11 15:57:25.412036323 +0000 UTC"


ArgoCD server logs 
time="2021-11-11T15:11:39Z" level=info msg="Initializing OIDC provider (issuer: https://172.28.8.5/api/dex)"
time="2021-11-11T15:11:41Z" level=info msg="received unary call /version.VersionService/Version" grpc.method=Version grpc.request.claims=null grpc.request.content= grpc.service=version.VersionService grpc.start_time="2021-11-11T15:11:41Z" span.kind=server system=grpc
time="2021-11-11T15:11:41Z" level=info msg="finished unary call with code OK" grpc.code=OK grpc.method=Version grpc.service=version.VersionService grpc.start_time="2021-11-11T15:11:41Z" grpc.time_ms=8.721 span.kind=server system=grpc
time="2021-11-11T15:11:41Z" level=info msg="received unary call /cluster.SettingsService/Get" grpc.method=Get grpc.request.claims=null grpc.request.content= grpc.service=cluster.SettingsService grpc.start_time="2021-11-11T15:11:41Z" span.kind=server system=grpc
time="2021-11-11T15:11:41Z" level=info msg="Ignore status for CustomResourceDefinitions"
time="2021-11-11T15:11:41Z" level=info msg="Ignore '/spec/preserveUnknownFields' for CustomResourceDefinitions"
time="2021-11-11T15:11:41Z" level=info msg="finished unary call with code OK" grpc.code=OK grpc.method=Get grpc.service=cluster.SettingsService grpc.start_time="2021-11-11T15:11:41Z" grpc.time_ms=3.787 span.kind=server system=grpc
time="2021-11-11T15:11:41Z" level=info msg="received unary call /cluster.SettingsService/Get" grpc.method=Get grpc.request.claims=null grpc.request.content= grpc.service=cluster.SettingsService grpc.start_time="2021-11-11T15:11:41Z" span.kind=server system=grpc
time="2021-11-11T15:11:41Z" level=info msg="Ignore status for CustomResourceDefinitions"
time="2021-11-11T15:11:41Z" level=info msg="Ignore '/spec/preserveUnknownFields' for CustomResourceDefinitions"
time="2021-11-11T15:11:41Z" level=info msg="finished unary call with code OK" grpc.code=OK grpc.method=Get grpc.service=cluster.SettingsService grpc.start_time="2021-11-11T15:11:41Z" grpc.time_ms=3.61 span.kind=server system=grpc
time="2021-11-11T15:11:41Z" level=info msg="received unary call /cluster.SettingsService/Get" grpc.method=Get grpc.request.claims=null grpc.request.content= grpc.service=cluster.SettingsService grpc.start_time="2021-11-11T15:11:41Z" span.kind=server system=grpc
time="2021-11-11T15:11:41Z" level=info msg="Ignore status for CustomResourceDefinitions"
time="2021-11-11T15:11:41Z" level=info msg="Ignore '/spec/preserveUnknownFields' for CustomResourceDefinitions"
time="2021-11-11T15:11:41Z" level=info msg="finished unary call with code OK" grpc.code=OK grpc.method=Get grpc.service=cluster.SettingsService grpc.start_time="2021-11-11T15:11:41Z" grpc.time_ms=3.95 span.kind=server system=grpc
time="2021-11-11T15:11:41Z" level=info msg="received unary call /session.SessionService/GetUserInfo" grpc.method=GetUserInfo grpc.request.claims=null grpc.request.content= grpc.service=session.SessionService grpc.start_time="2021-11-11T15:11:41Z" span.kind=server system=grpc
time="2021-11-11T15:11:41Z" level=info msg="finished unary call with code OK" grpc.code=OK grpc.method=GetUserInfo grpc.service=session.SessionService grpc.start_time="2021-11-11T15:11:41Z" grpc.time_ms=1.703 span.kind=server system=grpc
time="2021-11-11T15:11:43Z" level=info msg="Initializing OIDC provider (issuer: https://172.28.8.5/api/dex)"

**Comment by jannfis** [comment]
Moving this to discussions, because it's obviously not a bug.
